pulumi/infra/kafka.py

```diff
@@ -1,13 +1,137 @@
+
+
+@dataclasses.dataclass
+class ServiceOutput:
+    ingress_topics: Sequence[str]
+    egress_topics: Sequence[str]
+    service_account: CredentialOutput
+
+    @staticmethod
+    def from_json(json_: Mapping[str, Any]) -> ServiceOutput:
+        return ServiceOutput(
+            ingress_topics=json_["ingress_topics"],
+            egress_topics=json_["egress_topics"],
+            service_account=CredentialOutput.from_json(json_["service_account"]),
+        )
+
+
+@dataclasses.dataclass
+class EnvironmentOutput:
+    environment_id: str
+    bootstrap_servers: str
+    environment_credentials: CredentialOutput
+    services: Mapping[str, ServiceOutput]
+    topics: Mapping[str, TopicOutput]
+
+    def get_service_credentials(self, service_name: str) -> CredentialOutput:
+        if service_name in self.services:
+            return self.services[service_name].service_account
+        else:
+            raise KeyError(f"{service_name} does not exist")
+
+    @staticmethod
+    def from_json(json_: Mapping[str, Any]) -> EnvironmentOutput:
+        return EnvironmentOutput(
+            environment_id=json_["environment_id"],
+            bootstrap_servers=json_["bootstrap_servers"],
+            environment_credentials=CredentialOutput.from_json(
+                json_["environment_credentials"]
+            ),
+            services={
+                k: ServiceOutput.from_json(v) for k, v in json_["services"].items()
+            },
+            topics={k: TopicOutput.from_json(v) for k, v in json_["topics"].items()},
+        )
+
+
+@dataclasses.dataclass
+class ConfluentOutput:
+    environments: Mapping[str, EnvironmentOutput]
+
+    def get_environment(self, environment_name: str) -> EnvironmentOutput:
+        if environment_name in self.environments:
+            return self.environments[environment_name]
+        else:
+            raise KeyError(f"{environment_name} does not exist")
+
+    @staticmethod
+    def from_json(
+        json_: pulumi.Output[Mapping[str, Any]]
+    ) -> pulumi.Output[ConfluentOutput]:
+        return json_.apply(
+            lambda j: ConfluentOutput(
+                environments={k: EnvironmentOutput.from_json(v) for k, v in j.items()}
+            )
+        )
+
 
 class Kafka(pulumi.ComponentResource):
-    def __init__(self, name: str, opts: Optional[pulumi.ResourceOptions] = None):
+    confluent_environment: pulumi.Output[EnvironmentOutput]
+
+    def __init__(
+        self,
+        name: str,
+        confluent_environment_name: str,
+        opts: Optional[pulumi.ResourceOptions] = None,
+    ):
         super().__init__("grapl:Kafka", name=name, props=None, opts=opts)
-        metrics_topic = kafka.Topic(
-            "metrics-topic", name="metrics", replication_factor=1, partitions=1
-        )
+
+        confluent_stack_output = StackReference(
+            "grapl/ccloud-bootstrap/ccloud-bootstrap"
+        ).get_output("confluent")
+
+        assert confluent_stack_output is not None
```

Comment:
If you use [`require_output`](https://github.com/pulumi/pulumi/blob/722d8344e1013d8bf1e68f6394011e1fee12e2b2/sdk/python/lib/pulumi/stack_reference.py#L74-L85) instead of `get_output` on line 131 above, you could leave this assertion out.